src/Bedrock.Framework/Protocols/MessagePipeReader.cs

```diff
@@ -0,0 +1,155 @@
+using System.Collections.Generic;
+using System.IO;
+using System.IO.Pipelines;
+using System.Runtime.InteropServices;
+using System.Text;
+using System.Threading;
+using System.Threading.Tasks;
+
+namespace Bedrock.Framework.Protocols
+{
+    public class MessagePipeReader : PipeReader
+    {
+        private IMessageReader<ReadOnlySequence<byte>> _messageReader;
+        private readonly PipeReader _reader;
+
+        private SequencePosition _examined;
+        private SequencePosition _consumed;
+        private ReadOnlySequence<byte> _message;
+        private bool _isCanceled;
+        private bool _isCompleted;
+        private ConsumableArrayBufferWriter<byte> _backlog = new ConsumableArrayBufferWriter<byte>();
+        private bool _allExamined;
+
+        public MessagePipeReader(PipeReader reader, IMessageReader<ReadOnlySequence<byte>> messageReader)
+        {
+            _reader = reader ?? throw new ArgumentNullException(nameof(reader));
+            _messageReader = messageReader ?? throw new ArgumentNullException(nameof(messageReader));
+        }
+
+        public override void AdvanceTo(SequencePosition consumed)
+        {
+            AdvanceTo(consumed, consumed);
+        }
+
+        public override void AdvanceTo(SequencePosition consumed, SequencePosition examined)
+        {
+            if (examined.Equals(_message.End))
+            {
+                _allExamined = true;
+            }
+            
+            var consumedLength = (int)_message.Slice(_message.Start, consumed).Length;
+
+            if (_backlog.UnconsumedWrittenCount > 0)
+            {
+                _backlog.Consume(consumedLength);
+            }
+            else
+            {
+                var unconsumed = _message.Slice(consumed);
+                if (!unconsumed.IsEmpty)
+                {
+                    foreach (var m in unconsumed)
+                    {
+                        _backlog.Write(m.Span);
+                    }
+                }
+            }
+
+            // REVIEW: Use the correct value for examined
+            _reader.AdvanceTo(_consumed, _examined);
+        }
+
+        public override void CancelPendingRead()
+        {
+            _reader.CancelPendingRead();
+        }
+
+        public override void Complete(Exception exception = null)
+        {
+            _reader.Complete();
+        }
+
+        public override async ValueTask<ReadResult> ReadAsync(CancellationToken cancellationToken = default)
+        {
+            while (true)
+            {
+                var result = await _reader.ReadAsync(cancellationToken).ConfigureAwait(false);
+
+                if (TryCreateReadResult(result, out var readResult))
+                    return readResult;
+            }
+        }
+
+        public override bool TryRead(out ReadResult readResult)
+        {
+            if (_reader.TryRead(out var result))
+            {
+                if (TryCreateReadResult(result, out readResult))
```

Comment:
nit: style wise, no self closing braces allowed (I need to update editor.config)